src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,144 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        final Node<T> last = tail;
+        final Node<T> newNode = new Node<>(last, value, null);
+        tail = newNode;
+        checkIfHeadIsNull(newNode);
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index >= 0 && size == 0 || index == size) {
+            add(value);
+            return;
+        }
+        if (index >= 0 && index < size) {
+            Node<T> nextNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(nextNode.prev, value, nextNode);
+            checkIfHeadIsNull(newNode);
+            nextNode.prev = newNode;
+            size++;
+            return;
+        }
+        throw new IndexOutOfBoundsException("Index " + index + " is invalid!");
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldValue = getNodeByIndex(index).value;
+        getNodeByIndex(index).value = value;
```

Comment:
```suggestion
        Node<T> node = getNodeByIndex(index);
        T oldValue = node.value;
        node.value = value;
```
you are iterating 2 times